Changing logging to use logrus.WithFields

jterry75 opened the pull request:
Signed-off-by: Justin Terry (VM) <juterry@microsoft.com>

jterry75 commented:
Not ready for this to merge still have a lot of work to do in the other packages but before I go crazy I want some early feedback on the approach and if you think this will be helpful to do it in this way.

jterry75 commented on internal/hcs/watcher.go:
```diff
@@ -17,14 +18,16 @@ import (
 	}
-	logrus.Warnf("%s: Did not complete within %s. This may indicate a platform issue. If it appears to be making no forward progress, obtain the stacks and see is there is a syscall stuck in the platform API for a significant length of time.", description, timeout.SyscallWatcher)
+	logrus.WithFields(context).
+		WithField(logfields.Timeout, timeout.SyscallWatcher).
```
Bug. Should read `timeout.SycallWatcher.Nanoseconds()`

jterry75 commented:
@kevpar - FYI

jstarks commented on internal/copywithtimeout/copywithtimeout.go:
```diff
@@ -31,11 +31,10 @@ func init() {
+	logrus.WithField("stdval", context).
+		WithField("size", size).
+		WithField("timeout", timeout.Nanoseconds()).
+		Debug("hcsshim::copywithtimeout - Begin")
```
This doesn't seem like the kind of logging we aspire to...

jstarks replied:
In other words, this one seems quite verbose and also probably useless...

jterry75 replied:
It doesn't bother me either way. For one off's this seemed to do the trick. As you see in the process/system where we have an object with context we do the keys a different way so we do use the `WithFields`

This could equally be written
```go
fields := logrus.Fields{
    "stdval": context,
    "size": size,
    "timeout": timeout.Nanoseconds(),
}
logrus.WithFields(fields).Debug("hcsshim::copywithtimeout - Begin")
```

jterry75 replied:
Moved to the inline style for fields decl.

jstarks commented:
The example in https://godoc.org/github.com/sirupsen/logrus shows using `WithFields` for multi-field logging, not individual `WithField` calls. Which makes sense, because `WithFields` appears to be more efficient.

jstarks commented:
Also, `WithError` just adds a field called "error" (or alternatively, `logrus.ErrorKey`), so we don't need to use it separately if there are other fields.

jstarks commented on internal/hcs/process.go:
```diff
@@ -72,13 +72,45 @@ func (process *Process) SystemID() string {
+
+	operation := "hcsshim::Process::Signal"
+	process.logOperationBegin(operation)
+	defer func() {
```
This can be written as `defer process.logOperationEnd(operation, err)`.

You don't need the separate `func()`.

jstarks replied:
Unless this changes the capture rules for `err`?

jterry75 replied:
Yes good point. I originally put the error handler in there but now its not needed.

jstarks commented on internal/hcs/process.go:
```diff
@@ -190,22 +228,27 @@ func (process *Process) ResizeConsole(width, height uint16) error {
 
 	var resultp *uint16
+	completed := false
+	go syscallWatcher(process.logGetCommonFields(operation), &completed)
```
Seems like there's a functional change in here. Should this be a separate review?

jterry75 replied:
Yes. I will submit these in another review. It looks like @jhowardmsft missed a few with his `syscallWatcher` so I will PR these up as bugs

jterry75 replied:
#388

jstarks commented on internal/hcs/watcher.go:
```diff
@@ -17,14 +18,16 @@ import (
 	}
-	logrus.Warnf("%s: Did not complete within %s. This may indicate a platform issue. If it appears to be making no forward progress, obtain the stacks and see is there is a syscall stuck in the platform API for a significant length of time.", description, timeout.SyscallWatcher)
+	logrus.WithFields(context).
+		WithField(logfields.Timeout, timeout.SyscallWatcher.Nanoseconds()).
```
What happens if you just pass `timeout.SyscallWatcher`, without the `.Nanoseconds()`?

Regardless, having the log specify nanoseconds as an integer seems like it will be hard to read.

jterry75 replied:
Wouldn't we want the raw data to be in NS so that any viewer can display it in "pretty" print? But we don't want to drop the rounding early correct?

jterry75 replied:
Will very what the default for passing a `timeout` is though.

jstarks replied:
Well, you're throwing the type information away here, so e.g. if we build TraceLogging support for logrus, then we can't specifically log this as a duration.

So I guess the question is what does the JSON logger do with `time.Time`?

jterry75 commented:
@jstarks - I don't mind if we write `logrus.ErrorKey` into fields ourselves if we are already creating the `logrus.Fields` map previous to calling the log message.

In practice I feel like I want to write code like this (obviously more polished):

```go
func DoWork(ctx context.Context, ...) (err error) {
    ctx := log.Start(ctx, "DoWork")
    defer log.End(ctx, "DoWork")

    // Do something
    if err = ChildOperation(ctx); err != nil {
        return err
    }

    // ctx is restored to the original DoWork context
    SameContext(ctx)
}

func SubOperation(ctx context.Context) error {
    // Start a subcontext that tracks its parent but does not affect it.
    ctx := log.Start(ctx, "SubOperation")
    defer log.End(ctx, "SubOperation")

   id = "test" // RealWork that creates id
    ctx.WithFields("id", test)

    log.Debug(ctx, "Success")
    return nil
}

func SameContext(ctx context.Context) {
    ctx.WithFields("flows-up", true)
    log.Debug(ctx).Debug("SameContext - Success")
}
```
And then see output like this:
```bash
{ "msg": "Start Operation DoWork" }
{ "msg": "Start Operation SubOperation", "parent": "DoWork" }
{ "msg": "SubOperation: Success", "id": "test", "parent": "DoWork" }
{ "msg": "End Operation SubOperation", "id": "test", "parent": "DoWork" }
{ "msg": "DoWork: SameContext - Success", "flows-up": true }
{ "msg": "End Operation DoWork", "flows-up": true }
```

jstarks commented on cmd/runhcs/container.go:
```diff
@@ -573,7 +592,10 @@ func createContainerInHost(c *container, vm *uvm.UtilityVM) (err error) {
+	logrus.WithFields(logrus.Fields{
+		logfields.ContainerID: c.ID,
+		logfields.UVMID:       vmid,
+	}).Infof("creating container in UVM")
```
Should just be .Info

jstarks commented on internal/hcs/watcher.go:
```diff
@@ -17,14 +18,16 @@ import (
 // completed = true
 //
-func syscallWatcher(description string, syscallCompleted *bool) {
+func syscallWatcher(context logrus.Fields, syscallCompleted *bool) {
```
I wonder whether this should take a *logrus.Entry instead of a logrus.Fields, and the common field routine should return an entry instead of the field map. 

This is a minor difference, but it does remove one allocation. Also, it gives us flexibility to choose the non-standard logger without having to change this function.

jstarks commented on internal/logfields/fields.go:
```diff
@@ -0,0 +1,41 @@
+package logfields
+
+// Identifiers
+const (
```
nit: I think conventionally we would put these all in one big const block, with blank lines separating different categories.

jterry75 commented:
@jstarks @jhowardmsft - PTAL. I removed any additional `syscallWatcher` that I added. I will wait until #388 to fix these up if we decide to take that. I also changed `process` and `system` to use a statically created `logctx` field rather than alloc the map for each syscall begin/end which simplifies the maps nicely. I think this is good to go for at least these functions. Still need to convert `internal/uvm` and `internal/hcsoci` to this style for the best coverage.

jterry75 commented:
Merging this for now. Will add two additional changes for `internal/uvm` and `internal/hcsoci` logging changes.
